### ISSUE TITLE
Accessibility

### ISSUE THREAD

**Issue by DannyJi** [IcaliaLabs/Presentr#51]
I am working with a bunch of presentr's to show various controllers throughout my app.  I love it.

I am asked to do an Accessibility sweep in prep for use by government.  They test using VoiceOver.  When I use Voice Over, I am unable to dismiss the presented controller, by tapping outside.  Also when tapping and dragging on the blurred background behind, Voice Over is reading my underlying controller interface items.

I have tried using accessibilityViewIsModal with no luck.  tapToDismiss is on as well.  Without Voice Over on , all works as designed.

Has anyone had any experience with this?  We have to give the sight-mpaired a way to close the presented controller when voice over is on.

**Comment by DannyJi** [comment]
I was able to get  `accessibilityViewIsModal` to work properly.  Now Voice Over ignores UI elements in underlying view (below the blurred background in my case)
   
  ```
      presenter.blurBackground = true
      definitionsVC.view.accessibilityViewIsModal = true
      customPresentViewController(presenter, viewController: definitionsVC, animated: true, completion: nil)
```

I am still not having luck dismissing the controller when VoiceOver is on.  I may have to resort to adding a close button in the corner.  Hoping someone has another solution

**Comment by danlozano** [comment]
Hey @DannyJi !

Thanks for using Presentr, I'm glad you love it.

I don't have much experience with Accessibilty, I just assumed that everything would work out of the box. But this is something really important, and I'd be glad to look into it.

So, the first problem was that Voice Over was reading the views behind the blurred background, but you managed to fix that using the accesibilityViewIsModal property on the presented VC, correct?

Your other problem is that tapping on the background, does NOT dismiss the view controller? Do you have any sample code I can try out? Or can you tell me the steps necessary to recreate the problem myself? so I can try and see if there is a fix for that. For some reason the background view is NOT receiving touch I assume.

For now you could add a button do dismiss the VC, but anyways we should fix this.

👍

**Comment by DannyJi** [comment]
Hey @danlozano,

Thanks for your response.

1) Yes.  Adding accesibilityViewIsModal to the presented VC's **view** solved the problem where, without that VoiceOver would be "speaking" items behind the view.

2) Yes.  I cannot tap outside the presented view to dismiss the VC as normal when VoiceOver is active.  Without it, it works as expected.  I do not have sample code to try at this point (getting an app ready to send for testing tonight), but any app with a presented VC using Presentr and VoiceOver enable should demonstrate the problem.  You will need to try this on a real device, enable VoiceOver from the Settings > General > Accessibility.  Note that VoiceOver will change your tap pattern, for instance most elements you need to double tap where you would normally tap.

I think your assessment is correct.. the background not receiving the touch.

I will try to work on this in a few hours myself after some deployment meetings.

Thanks again.
Danny

**Comment by DannyJi** [comment]
Hey @danlozano,

For the second issue (tapping the background does not dismiss the presented view controller), I first started as we discussed - adding a close button begrudgingly.  While I got this to work, I was not happy with the way it looked and the code I needed to copy around.

I did some more research and found a great article [Adding Accessible Behavior](http://ronnqvi.st/adding-accessible-behavior/)

There are two "global gestures" or concepts that can be used to dismiss a modal window.  One is a magic tap, but the other is called the escape gesture.  That is the one I wanted.  The user can simply draw a "Z" with 2 fingers on the touch surface.  It is an escape functionality that takes you back a step when VoiceOver is activated.  I just needed to implement the `accessibilityPerformEscape()` method on the ViewControllers that would be presented in the app by Presentr.

```
 override func accessibilityPerformEscape() -> Bool {
    self.dismiss(animated: true, completion: nil)
    return true
  }
```

This worked, was a lot less coding and helped pass an accessibility review.

It would be nice if these could be incorporated into Presentr as an option somehow - maybe in its configuration.

**Possible Future Enhancement**
```
.
.
presenter.blurBackground = true
presenter.accessibilityPresentedViewIsModal = true
presenter.accessibilityEnableEscapeToDismiss = true
.
```

I just wanted to share back.

**Comment by danlozano** [comment]
Hey @DannyJi !

Thanks for the help, I've been really busy.

Let me see if there is a way to integrate that functionality inside Presentr, it would be great!

Thanks for doing the research :) I'lll get back to you soon.

**Comment by danlozano** [comment]
Hi @DannyJi !

Thanks for sharing the info. I've been really busy at work lately so I have not had time to update or look into Presentr's issues. Finally have some days off to work solely on my open source projects so I'm looking into everything.

So basically for voiceover to ignore the elements under the modal you have to set
view.accesibilityViewIsModal on the view controller.

And if you want to be able to dismiss the view controller you have to override the
accesibilityPerformEscape on the view controller.

After giving it some thought I realized that this is something outside the responsibilities of Presentr. These are "customizations" you are doing on your own view controller to make it behave "differently" or in this case, work properly.

Besides, there is no easy/non-hacky way to automatically insert the overridden method into the presented view controller by Presentr. I think that is better to be left to the developer.

What I could do to make use of your research in finding this "bug" is add a section in the Readme under accesibilty, and outline the things someone might want to do on their own view controller if they are in a similar case.

Sound good?

👍
